### ISSUE TITLE
Clustering doesn't work in ipv6 only environment

### ISSUE THREAD

**Issue by mariia-zelenova** [rabbitmq/rabbitmq-server#1421]
In 3.6.10 the following arguments make clustering work:
```
SERVER_ERL_ARGS="-proto_dist inet6_tcp +P 1048576"
CTL_ERL_ARGS="-proto_dist inet6_tcp +P 1048576"
```
In 3.6.13 command ```rabbitmqctl join_cluster rabbit@hostname2``` on hostname1 with arguments
```
SERVER_ADDITIONAL_ERL_ARGS="-proto_dist inet6_tcp"
```
causes ```unable to connect to epmd (port 4369) on hostname2: nxdomain (non-existing domain)```
and
```
SERVER_ADDITIONAL_ERL_ARGS="-proto_dist inet6_tcp"
CTL_ERL_ARGS="-proto_dist inet6_tcp +P 1048576"
``` 
causes ```TCP connection succeeded but Erlang distribution failed``` while connecting to hostname1.

**Comment by michaelklishin** [comment]
Thank you for your time.

Team RabbitMQ uses GitHub issues for specific actionable items engineers can work on. This assumes two things:

1. GitHub issues are not used for questions, investigations, root cause analysis, discussions of potential issues, etc (as defined by this team)
2. We have a certain amount of information to work with

We get at least a dozen of questions through various venues every single day, often quite light on details.
At that rate GitHub issues can very quickly turn into a something impossible to navigate and make sense of even for our team. Because of that questions, investigations, root cause analysis, discussions of potential features are all considered to be mailing list material by our team. Please post this to [rabbitmq-users](https://groups.google.com/forum/#!forum/rabbitmq-users).

Getting all the details necessary to reproduce an issue, make a conclusion or even form a hypothesis about what's happening can take a fair amount of time. Our team is multiple orders of magnitude smaller than the RabbitMQ community. Please help others help you by providing a way to reproduce the behavior you're
observing, or at least sharing as much relevant information as possible [on the list](https://groups.google.com/forum/#!forum/rabbitmq-users):

 * Server, client library and plugin (if applicable) versions used
 * Server logs
 * A code example or terminal transcript that can be used to reproduce
 * **Full** exception stack traces (not a single line message)
 * `rabbitmqctl status` (and, if possible, `rabbitmqctl environment` output)
 * Other relevant things about the environment and workload, e.g. [a traffic capture](https://www.rabbitmq.com/amqp-wireshark.html)

Feel free to edit out hostnames and other potentially sensitive information.

When/if we have enough details and evidence we'd be happy to file a new issue.

Thank you.

**Comment by michaelklishin** [comment]
@mariia-zelenova neither error message has anything to with IPv6.

> nxdomain (non-existing domain)

means that `hostname2` fails to resolve, presumably because it doesn't have an AAAA record.

> TCP connection succeeded but Erlang distribution failed

is covered both [in the docs](http://www.rabbitmq.com/clustering.html) and [in rabbitmq-users archives](https://groups.google.com/d/msg/rabbitmq-users/a6sqrAUX_Fg/SHt00K3sAwAJ).

**Comment by michaelklishin** [comment]
On top of the above, `SERVER_ERL_ARGS` and `SERVER_ADDITIONAL_ERL_ARGS` are not the same thing: the former *replaces* Erlang VM command line arguments while the latter *appends* to the default set. We definitely recommend using the latter but the point is that you are not comparing apples to oranges.

**Comment by zhangguanzhang** [comment]
I finally solved the problem by myself, and since I didn't find a place to share my experience, I replied here.

- ensure [rabbitmq version >= 3.6.3](https://bugzilla.redhat.com/show_bug.cgi?id=1358311)

- and the [distribution-ipv6](https://www.rabbitmq.com/docs/networking#distribution-ipv6) cannot work , adding them into file `rabbitmq-env.conf`  insteadof environment:
```
$ cat /etc/rabbitmq/rabbitmq-env.conf
# these flags will be used by RabbitMQ nodes
RABBITMQ_SERVER_ADDITIONAL_ERL_ARGS="-proto_dist inet6_tcp"
# these flags will be used by CLI tools
RABBITMQ_CTL_ERL_ARGS="-proto_dist inet6_tcp"
```

- Make sure that the hostname in the nodename `<prefix>@<hostname>` format, and `hostname` can be resolved. You can use the following command to test:
```
$ rabbitmq-diagnostics resolve_hostname rabbitmq-cluster1 --address-family IPv6
Asking node rabbitmq-cluster2@rabbitmq-cluster2 to resolve hostname rabbitmq-cluster1 to IPv6 addresses...
2408:8656:22df:ff01::13:181

$ getent hosts rabbitmq-cluster1
2408:8656:22df:ff01::13:181 rabbitmq-cluster1
```

- Similarly, the IP address of the network card to be monitored needs to be modified in the configuration file.

```
$ cat rabbitmq.conf
...
distribution.listener.interface = ::
listeners.tcp.default = 2408:8656:22df:ff01::13:182:5672
listeners.tcp.management = 2408:8656:22df:ff01::13:182:15672 
```

- Also make sure that your nodename is in the `<prefix>@<hostname>` format and not just the `<hostname>` part. If you do not use a configuration file to configure nodename, use the environment variable `RABBITMQ_NODENAME`

**Comment by michaelklishin** [comment]
@zhangguanzhang you can start a new discussion and explain that you'd like to share a solution because the original discussion was closed (the version used was `3.12.x`). There's nothing wrong with that.

Great to see someone use `rabbitmq-diagnostics resolve_hostname`. It was added to simplify troubleshooting and it looks like it did come handy in your case.

**Comment by zhangguanzhang** [comment]
> [@zhangguanzhang](https://github.com/zhangguanzhang) you can start a new discussion and explain that you'd like to share a solution because the original discussion was closed (the version used was `3.12.x`). There's nothing wrong with that.
> 
> Great to see someone use `rabbitmq-diagnostics resolve_hostname`. It was added to simplify troubleshooting and it looks like it did come handy in your case.

Yesterday, the discussion I submitted was directly locked and I couldn't reply. This issue is the top link on Google when searching for "rabbitmq ipv6" , so I'm replying here to make it easier for others to find this through search engines in the future.
